tripal/src/Services/TripalEntityLookup.php

```diff
@@ -0,0 +1,260 @@
+    $bundle_id = NULL;
+    $bundles = \Drupal::entityTypeManager()
+      ->getStorage('tripal_entity_type')
+      ->loadByProperties(['termIdSpace' => $termIdSpace, 'termAccession' => $termAccession]);
+    if (sizeof($bundles) == 1) {
+      $bundle_id = key($bundles);
+    }
+    return $bundle_id;
+  }
+
+  /**
+   * Retrieve the base table for a given bundle.
+   *
+   * @param string $bundle
+   *   The bundle's ID, e.g. "gene"
+   * @param string $entity_type
+   *   The type of entity, only 'tripal_entity' is supported.
+   *
+   * @return string|null
+   *   The base table name, or null if no match found.
+   */
+  protected function getBundleBaseTable($bundle, $entity_type) {
+    $table = NULL;
+    if ($bundle) {
+      $entityFieldManager = \Drupal::service('entity_field.manager');
+      $fields = $entityFieldManager->getFieldDefinitions($entity_type, $bundle);
+      $field_list = array_keys($fields);
+      $type_name = NULL;
+      $base_table = NULL;
+      foreach ($field_list as $field_name) {
+        // Skip drupal fields. Look for the first tripal field that has a base_column
+        // set. Fields from linker tables do not have a base_column.
+        if (preg_match('/^'.$bundle.'/', $field_name)) {
+          $field_storage = FieldStorageConfig::loadByName($entity_type, $field_name);
+          if ($field_storage) {
+            $storage_plugin_settings = $field_storage->getSettings()['storage_plugin_settings'];
+            $base_table = $storage_plugin_settings['base_table'];
+            $base_column = $storage_plugin_settings['base_column'] ?? '';
+            if ($base_table and $base_column) {
+              $table = $base_table;
+              break;
+            }
+          }
+        }
+      }
+    }
+    return $table;
+  }
+
+  /**
+   * Retrieve the pkey for an entity corresponding to a given record in a given table.
+   *
+   * @param string $base_table
+   *   The name of the chado table
+   * @param int $record_id
+   *   The primary key value for the requested record in the $base_table
+   * @param string $bundle_id
+   *   The name of the drupal bundle, e.g. for base table 'arraydesign' it is 'array_design'
+   * @param string $entity_type
+   *   The type of entity, only 'tripal_entity' is supported.
+   *
+   * @return array
+   *   The id for the requested Drupal entity in the tripal_entity table.
+   *   Will be an empty array if there was no corresponding entity. This can happen
+   *   if the content is not published, or if this is not a content type.
+   *   If a given record was published as more than one content type, the
+   *   returned array may have more than one entity id. This would happen
+   *   in Tripal 3, for example if an analysis is published as both
+   *   "analysis", and as "Genome Assembly".
+   */
+  protected function getEntityIdFromRecordId($base_table, $record_id, $bundle_id, $entity_type) : array {
```

Comment:
I was worrying about the assuming the field database table names and fell down a rabbit hole 🙈 
Apparently we can use the entity query api to search for all entities with a particular field value. Furthermore, we can dig into the properties of that field!

In the following example, I am grabbing the entity_id of all tripal entities with a bundle of contact and a contact_name field whose record_id property has a value of 2.
```php
$query = \Drupal::entityQuery('tripal_entity')
  ->condition('type', 'contact')
  ->condition('contact_name.record_id', 2)
  ->accessCheck(TRUE);
$results = $query->execute();
```

This can be used to get rid of the need to know the base table and to remove the assumptions regarding database tables!
```php
$query = \Drupal::entityQuery('tripal_entity')
  ->condition('type', $bundle_id)
  ->condition($required_field . '.record_id', $record_id)
  ->accessCheck(TRUE);
$results = $query->execute();
```
This does still make the assumption that we named the primary key of the base table `record_id` 🙈

Reply:
This assumption will be dealt with in issue #1850